ppa/archive/views.py

```diff
@@ -166,3 +170,73 @@ def render_to_csv(self, data):
 
     def get(self, *args, **kwargs):
         return self.render_to_csv(self.get_data())
+
+
+class BulkAddCollectionView(ListView, FormMixin, ProcessFormView):
+    '''
+    View to bulk add a queryset of :class:ppa.archive.models.DigitizedWork
+    to a set of :class:ppa.archive.models.Collection instances.
+
+    Restricted to staff users via staff_member_required on url.
+    '''
+
+    model = DigitizedWork
+    template_name = 'admin/archive/digitizedwork/bulk_add_collections.html'
+    form_class = BulkAddCollectionForm
+    success_url = reverse_lazy('admin:archive_digitizedwork_changelist')
+
+    def get_queryset(self, *args, **kwargs):
+        '''Return a queryset filtered by id, or empty list if no ids'''
+        ids = self.request.GET.get('ids', None)
+        # if somehow a problematic non-pk is pushed, will be ignored in filter
+        return DigitizedWork.objects.filter(id__in=ids.split(',')
+                                            if ids else [])
+
+    def get_initial(self):
+        '''
+        Set the hidden field with the :class:ppa.archive.models.DigitizedWork
+        to add to collection(s) to pass to POST request.
+        '''
+        initial = super(BulkAddCollectionView, self).get_initial()
+        # this ensures that the hidden field for POST only ever contains
+        # actual PKs
+        ids = ','.join(str(val) for val in
+                       self.get_queryset().values_list('id', flat=True))
+        if ids:
+            initial['digitized_work_ids'] = ids
+        return initial
+
+    def post(self, request, *args, **kwargs):
+        '''
+        Add :class:ppa.archive.models.DigitizedWork instances passed in form
+        data to selected instances of :class:ppa.archive.models.Collection,
+        then return to change_list view.
+        '''
+        form = BulkAddCollectionForm(request.POST)
+        # override default post functionality to set collections
+        if form.is_valid():
+            data = form.cleaned_data
+            ids = data['digitized_work_ids'].split(',')
+            # get digitzed works from validated form
+            digitized_works = DigitizedWork.objects.filter(id__in=ids)
+            for collection in data['collections']:
+                collection.digitizedwork_set.set(digitized_works)
+            # reindex solr with the new collection data
+            solr_docs = [work.index_data() for work in digitized_works]
+            solr, solr_collection = get_solr_connection()
+            solr.index(solr_collection, solr_docs,
+                params={'commitWithin': 2000})
+            # create a success message to add to message framework stating
+            # what happened
+            htids = ', '.join(work.source_id for work in digitized_works)
+            collections = ', '.join(collection.name for
+                                    collection in data['collections'])
+            messages.success(request, '%s successfully added to: %s.'
```

Comment:
I tried testing this with 100 works selected - the form works fine, but the message fails because it is too long.  Maybe just report the total number of works added to which collections?

Reply:
I revised to do just this. Numbers and collections added to are now noted.